### ISSUE TITLE
Support for C-style string arguments

### ISSUE THREAD

**Issue by jonhoo** [cuviper/rust-libprobe#10]
While all probe arguments do have to be `i64`, it seems like addresses to null-terminated strings are supported both by [the kernel](https://www.kernel.org/doc/html/latest/trace/uprobetracer.html#types) and by [`bpftrace`](https://github.com/iovisor/bpftrace/pull/106). We may want to make users' lives a little easier by also accepting [`CStr`](https://doc.rust-lang.org/std/ffi/struct.CStr.html) arguments and turning them into their own address.

**Comment by cuviper** [comment]
I could see this as some kind of `ProbeArg` trait which casts as necessary -- `CStr` would just need to call `as_ptr()`. It might even be possible to add an associated `const` for the sign+size, so we don't have to force everything to `i64`. See also:
https://github.com/cuviper/rust-libprobe/blob/431ac2999eb88e3a8ba5ee15df13557e234d9775/src/platform/systemtap.rs#L33-L54

**Comment by Zheaoli** [comment]
So, should we pass the str or String in Rust to the probe right now?

**Comment by cuviper** [comment]
Yeah, right now it will work with any argument that can cast `as isize` -- so `CString` and `CStr` won't work directly, but their `as_ptr()` should be fine. You could do that with `str` and `String` too, or any other slice, but you'll probably want to pass another argument for the length in that case.

**Comment by bobrik** [comment]
A native support via something like `ProbeArg` would indeed be nice. Currently with `probe_lazy` the `CString` created lazily are dropped and scrambled before they can be read:

```rust
probe_lazy!(proxy, h1, {
    let uri = CString::new("/lol.txt").unwrap();
    uri.as_ptr()
});
```

```
$ sudo bpftrace -e 'usdt:target/debug/examples/load_balancer:proxy:h1 { printf("uri = %s\n", str(arg0)); }'
Attaching 1 probe...
uri = �d�s
```

This happens with both `probe!` and `probe_lazy!` and can be avoided by creating the `CString` before a call to the probe, but that defies the purpose of lazy evaluation:

```rust
let uri = CString::new("/lol.txt").unwrap();
probe_lazy!(proxy, h1, { uri.as_ptr() });
```

```
$ sudo bpftrace -e 'usdt:target/debug/examples/load_balancer:proxy:h1 { printf("uri = %s\n", str(arg0)); }'
Attaching 1 probe...
uri = /lol.txt
```

**Comment by cuviper** [comment]
I think you can do this and still be lazy:

```rust
let uri;
probe_lazy!(proxy, h1, {
    uri = CString::new("/lol.txt").unwrap();
    uri.as_ptr()
});
```

It doesn't even need `mut` because there's only one (conditional) initialization.

**Comment by bobrik** [comment]
Indeed that does the trick, thanks!
